// minterjs-util/dist/cjs/variables.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.PAYLOAD_MAX_LENGTH = exports.MULTISIG_SIGNATURE_MAX_COUNT = exports.MULTISEND_RECIPIENT_MAX_COUNT = exports.COIN_MIN_MAX_SUPPLY = exports.COIN_MAX_MAX_SUPPLY = exports.COIN_MAX_AMOUNT = void 0;
/**
 * @deprecated
 * Maximum amount of any coin
 * @type {number}
 */
var COIN_MAX_AMOUNT = Math.pow(10, 15);
/**
 * Coin's maximum `maxSupply` value (in BIP)
 * @type {number}
 */
exports.COIN_MAX_AMOUNT = COIN_MAX_AMOUNT;
var COIN_MAX_MAX_SUPPLY = Math.pow(10, 15);
/**
 * Coin's minimum `maxSupply` value (in BIP)
 * @type {number}
 */
exports.COIN_MAX_MAX_SUPPLY = COIN_MAX_MAX_SUPPLY;
var COIN_MIN_MAX_SUPPLY = 1;

/**
 * Maximum number of signatures for multisig tx
 * @type {number}
 */
exports.COIN_MIN_MAX_SUPPLY = COIN_MIN_MAX_SUPPLY;
var MULTISIG_SIGNATURE_MAX_COUNT = 32;

/**
 * Maximum number of recipients of multisend tx
 * @type {number}
 */
exports.MULTISIG_SIGNATURE_MAX_COUNT = MULTISIG_SIGNATURE_MAX_COUNT;
var MULTISEND_RECIPIENT_MAX_COUNT = 100;

/**
 * Max length of tx payload (in bytes)
 * @type {number}
 */
exports.MULTISEND_RECIPIENT_MAX_COUNT = MULTISEND_RECIPIENT_MAX_COUNT;
var PAYLOAD_MAX_LENGTH = 10000;
exports.PAYLOAD_MAX_LENGTH = PAYLOAD_MAX_LENGTH;
