// minter-js-sdk/dist/cjs/tx-data/candidate-edit-public-key.js
'use strict';

var minterjsTx = require('minterjs-tx');
var minterjsUtil = require('minterjs-util');
var utils = require('../utils.js');

/**
 * @param {object} txData
 * @param {string} txData.publicKey
 * @param {string} txData.newPublicKey
 * @param {TxOptions} [options]
 * @constructor
 */
function EditCandidatePublicKeyTxData(_ref) {
  var publicKey = _ref.publicKey,
    newPublicKey = _ref.newPublicKey;
  var options = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
  if (!options.disableValidation) {
    utils.validatePublicKey(publicKey, 'publicKey');
    utils.validatePublicKey(newPublicKey, 'newPublicKey');
  }
  this.publicKey = publicKey;
  this.newPublicKey = newPublicKey;
  this.txData = new minterjsTx.TxDataEditCandidatePublicKey({
    publicKey: minterjsUtil.toBuffer(publicKey),
    newPublicKey: minterjsUtil.toBuffer(newPublicKey)
  });
  utils.proxyNestedTxData(this);
}

/**
 * @param {object} txData
 * @param {Buffer|string} txData.publicKey
 * @param {Buffer|string} txData.newPublicKey
 * @param {TxOptions} [options]
 * @return {EditCandidatePublicKeyTxData}
 */
EditCandidatePublicKeyTxData.fromBufferFields = function fromBufferFields(_ref2) {
  var publicKey = _ref2.publicKey,
    newPublicKey = _ref2.newPublicKey;
  var options = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
  return new EditCandidatePublicKeyTxData({
    publicKey: utils.dataToPublicKey(publicKey),
    newPublicKey: utils.dataToPublicKey(newPublicKey)
  }, options);
};

/**
 * @param {Buffer|string} data
 * @return {EditCandidatePublicKeyTxData}
 */
EditCandidatePublicKeyTxData.fromRlp = function fromRlp(data) {
  return EditCandidatePublicKeyTxData.fromBufferFields(new minterjsTx.TxDataEditCandidatePublicKey(data));
};

module.exports = EditCandidatePublicKeyTxData;
