// minter-js-sdk/dist/cjs/tx-data/candidate-edit-commission.js
'use strict';

var minterjsTx = require('minterjs-tx');
var minterjsUtil = require('minterjs-util');
var utils = require('../utils.js');

/**
 * @param {object} txData
 * @param {string} txData.publicKey
 * @param {number|string} txData.commission
 * @param {TxOptions} [options]
 * @constructor
 */
function EditCandidateCommissionTxData(_ref) {
  var publicKey = _ref.publicKey,
    commission = _ref.commission;
  var options = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
  if (!options.disableValidation) {
    utils.validatePublicKey(publicKey, 'publicKey');
    utils.validateUint(commission, 'commission');
  }
  this.publicKey = publicKey;
  this.commission = commission;
  this.txData = new minterjsTx.TxDataEditCandidateCommission({
    publicKey: minterjsUtil.toBuffer(publicKey),
    commission: utils.integerToHexString(commission)
  });
  utils.proxyNestedTxData(this);
}

/**
 * @param {object} txData
 * @param {Buffer|string} txData.publicKey
 * @param {Buffer|string} txData.commission
 * @param {TxOptions} [options]
 * @return {EditCandidateCommissionTxData}
 */
EditCandidateCommissionTxData.fromBufferFields = function fromBufferFields(_ref2) {
  var publicKey = _ref2.publicKey,
    commission = _ref2.commission;
  var options = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
  return new EditCandidateCommissionTxData({
    publicKey: utils.dataToPublicKey(publicKey),
    commission: utils.dataToInteger(commission)
  }, options);
};

/**
 * @param {Buffer|string} data
 * @return {EditCandidateCommissionTxData}
 */
EditCandidateCommissionTxData.fromRlp = function fromRlp(data) {
  return EditCandidateCommissionTxData.fromBufferFields(new minterjsTx.TxDataEditCandidateCommission(data));
};

module.exports = EditCandidateCommissionTxData;
