// minter-js-sdk/dist/cjs/tx-data/convert-sell-all.js
'use strict';

var minterjsTx = require('minterjs-tx');
var minterjsUtil = require('minterjs-util');
var utils = require('../utils.js');

// import {convertToPip} from 'minterjs-util/src/converter.js';

/**
 * @param {object} txData
 * @param {number|string} txData.coinToSell - coin id
 * @param {number|string} txData.coinToBuy - coin id
 * @param {number|string} [txData.minimumValueToBuy=0]
 * @param {TxOptions} [options]
 * @constructor
 */
function SellAllTxData(_ref) {
  var coinToSell = _ref.coinToSell,
    coinToBuy = _ref.coinToBuy,
    _ref$minimumValueToBu = _ref.minimumValueToBuy,
    minimumValueToBuy = _ref$minimumValueToBu === void 0 ? 0 : _ref$minimumValueToBu;
  var options = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
  if (!options.disableValidation) {
    utils.validateUint(coinToSell, 'coinToSell');
    utils.validateUint(coinToBuy, 'coinToBuy');
    utils.validateAmount(minimumValueToBuy, 'minimumValueToBuy');
  }
  this.coinToSell = coinToSell;
  this.coinToBuy = coinToBuy;
  this.minimumValueToBuy = minimumValueToBuy;
  this.txData = new minterjsTx.TxDataSellAll({
    coinToSell: utils.integerToHexString(coinToSell),
    coinToBuy: utils.integerToHexString(coinToBuy),
    minimumValueToBuy: "0x".concat(minterjsUtil.convertToPip(minimumValueToBuy, 'hex'))
  });
  utils.proxyNestedTxData(this);
}

/**
 * @param {object} txData
 * @param {Buffer|string} txData.coinToSell
 * @param {Buffer|string} txData.coinToBuy
 * @param {Buffer|string} txData.minimumValueToBuy
 * @param {TxOptions} [options]
 * @return {SellAllTxData}
 */
SellAllTxData.fromBufferFields = function fromBufferFields(_ref2) {
  var coinToSell = _ref2.coinToSell,
    coinToBuy = _ref2.coinToBuy,
    minimumValueToBuy = _ref2.minimumValueToBuy;
  var options = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
  return new SellAllTxData({
    coinToSell: utils.dataToInteger(coinToSell),
    coinToBuy: utils.dataToInteger(coinToBuy),
    minimumValueToBuy: utils.dataPipToAmount(minimumValueToBuy)
  }, options);
};

/**
 * @param {Buffer|string} data
 * @return {SellAllTxData}
 */
SellAllTxData.fromRlp = function fromRlp(data) {
  return SellAllTxData.fromBufferFields(new minterjsTx.TxDataSellAll(data));
};

module.exports = SellAllTxData;
