// minter-js-sdk/dist/cjs/tx-data/convert-buy.js
'use strict';

var minterjsTx = require('minterjs-tx');
var minterjsUtil = require('minterjs-util');
var utils = require('../utils.js');

// import {convertToPip} from 'minterjs-util/src/converter.js';

/**
 * @param {object} txData
 * @param {number|string} txData.coinToSell - coin id
 * @param {number|string} txData.coinToBuy - coin id
 * @param {number|string} txData.valueToBuy
 * @param {number|string} [txData.maximumValueToSell]
 * @param {TxOptions} [options]
 * @constructor
 */
function BuyTxData(_ref) {
  var coinToSell = _ref.coinToSell,
    coinToBuy = _ref.coinToBuy,
    valueToBuy = _ref.valueToBuy,
    _ref$maximumValueToSe = _ref.maximumValueToSell,
    maximumValueToSell = _ref$maximumValueToSe === void 0 ? minterjsUtil.COIN_MAX_AMOUNT : _ref$maximumValueToSe;
  var options = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
  if (!options.disableValidation) {
    utils.validateUint(coinToSell, 'coinToSell');
    utils.validateUint(coinToBuy, 'coinToBuy');
    utils.validateAmount(valueToBuy, 'valueToBuy');
    utils.validateAmount(maximumValueToSell, 'maximumValueToSell');
  }
  this.coinToSell = coinToSell;
  this.coinToBuy = coinToBuy;
  this.valueToBuy = valueToBuy;
  this.maximumValueToSell = maximumValueToSell;
  this.txData = new minterjsTx.TxDataBuy({
    coinToSell: utils.integerToHexString(coinToSell),
    coinToBuy: utils.integerToHexString(coinToBuy),
    valueToBuy: "0x".concat(minterjsUtil.convertToPip(valueToBuy, 'hex')),
    maximumValueToSell: "0x".concat(minterjsUtil.convertToPip(maximumValueToSell, 'hex'))
  });
  utils.proxyNestedTxData(this);
}

/**
 * @param {object} txData
 * @param {Buffer|string} txData.coinToSell
 * @param {Buffer|string} txData.valueToBuy
 * @param {Buffer|string} txData.coinToBuy
 * @param {Buffer|string} txData.maximumValueToSell
 * @param {TxOptions} [options]
 * @return {BuyTxData}
 */
BuyTxData.fromBufferFields = function fromBufferFields(_ref2) {
  var coinToSell = _ref2.coinToSell,
    valueToBuy = _ref2.valueToBuy,
    coinToBuy = _ref2.coinToBuy,
    maximumValueToSell = _ref2.maximumValueToSell;
  var options = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
  return new BuyTxData({
    coinToSell: utils.dataToInteger(coinToSell),
    coinToBuy: utils.dataToInteger(coinToBuy),
    valueToBuy: utils.dataPipToAmount(valueToBuy),
    maximumValueToSell: utils.dataPipToAmount(maximumValueToSell)
  }, options);
};

/**
 * @param {Buffer|string} data
 * @return {BuyTxData}
 */
BuyTxData.fromRlp = function fromRlp(data) {
  return BuyTxData.fromBufferFields(new minterjsTx.TxDataBuy(data));
};

module.exports = BuyTxData;
