// minter-js-sdk/dist/cjs/tx-data/convert-sell.js
'use strict';

var minterjsTx = require('minterjs-tx');
var minterjsUtil = require('minterjs-util');
var utils = require('../utils.js');

/**
 * @param {object} txData
 * @param {number|string} txData.coinToSell - coin id
 * @param {number|string} txData.coinToBuy - coin id
 * @param {number|string} txData.valueToSell
 * @param {number|string} [txData.minimumValueToBuy=0]
 * @param {TxOptions} [options]
 * @constructor
 */
function SellTxData(_ref) {
  var coinToSell = _ref.coinToSell,
    coinToBuy = _ref.coinToBuy,
    valueToSell = _ref.valueToSell,
    _ref$minimumValueToBu = _ref.minimumValueToBuy,
    minimumValueToBuy = _ref$minimumValueToBu === void 0 ? 0 : _ref$minimumValueToBu;
  var options = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
  if (!options.disableValidation) {
    utils.validateUint(coinToSell, 'coinToSell');
    utils.validateUint(coinToBuy, 'coinToBuy');
    utils.validateAmount(valueToSell, 'valueToSell');
    utils.validateAmount(minimumValueToBuy, 'minimumValueToBuy');
  }
  this.coinToSell = coinToSell;
  this.coinToBuy = coinToBuy;
  this.valueToSell = valueToSell;
  this.minimumValueToBuy = minimumValueToBuy;
  this.txData = new minterjsTx.TxDataSell({
    coinToSell: utils.integerToHexString(coinToSell),
    coinToBuy: utils.integerToHexString(coinToBuy),
    valueToSell: "0x".concat(minterjsUtil.convertToPip(valueToSell, 'hex')),
    minimumValueToBuy: "0x".concat(minterjsUtil.convertToPip(minimumValueToBuy, 'hex'))
  });
  utils.proxyNestedTxData(this);
}

/**
 * @param {object} txData
 * @param {Buffer|string} txData.coinToSell
 * @param {Buffer|string} txData.valueToSell
 * @param {Buffer|string} txData.coinToBuy
 * @param {Buffer|string} txData.minimumValueToBuy
 * @param {TxOptions} [options]
 * @return {SellTxData}
 */
SellTxData.fromBufferFields = function fromBufferFields(_ref2) {
  var coinToSell = _ref2.coinToSell,
    valueToSell = _ref2.valueToSell,
    coinToBuy = _ref2.coinToBuy,
    minimumValueToBuy = _ref2.minimumValueToBuy;
  var options = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
  return new SellTxData({
    coinToSell: utils.dataToInteger(coinToSell),
    coinToBuy: utils.dataToInteger(coinToBuy),
    valueToSell: utils.dataPipToAmount(valueToSell),
    minimumValueToBuy: utils.dataPipToAmount(minimumValueToBuy)
  }, options);
};

/**
 * @param {Buffer|string} data
 * @return {SellTxData}
 */
SellTxData.fromRlp = function fromRlp(data) {
  return SellTxData.fromBufferFields(new minterjsTx.TxDataSell(data));
};

module.exports = SellTxData;
