// minter-js-sdk/dist/cjs/tx-data/edit-ticker-owner.js
'use strict';

var minterjsTx = require('minterjs-tx');
var minterjsUtil = require('minterjs-util');
var utils = require('../utils.js');

/**
 * @param {object} txData
 * @param {string} txData.symbol
 * @param {string} txData.newOwner
 * @param {TxOptions} [options]
 * @constructor
 */
function EditTickerOwnerTxData(_ref) {
  var symbol = _ref.symbol,
    newOwner = _ref.newOwner;
  var options = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
  if (!options.disableValidation) {
    utils.validateTicker(symbol, 'symbol');
    utils.validateAddress(newOwner, 'newOwner');
  }
  this.symbol = symbol;
  this.newOwner = newOwner;
  this.txData = new minterjsTx.TxDataEditTickerOwner({
    symbol: minterjsUtil.coinToBuffer(symbol),
    newOwner: minterjsUtil.toBuffer(newOwner)
  });
  utils.proxyNestedTxData(this);
}

/**
 * @param {object} txData
 * @param {Buffer|string} txData.symbol
 * @param {Buffer|string} txData.newOwner
 * @param {TxOptions} [options]
 * @return {EditTickerOwnerTxData}
 */
EditTickerOwnerTxData.fromBufferFields = function fromBufferFields(_ref2) {
  var symbol = _ref2.symbol,
    newOwner = _ref2.newOwner;
  var options = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
  return new EditTickerOwnerTxData({
    symbol: minterjsUtil.bufferToCoin(minterjsUtil.toBuffer(symbol)),
    newOwner: utils.dataToAddress(newOwner)
  }, options);
};

/**
 * @param {Buffer|string} data
 * @return {EditTickerOwnerTxData}
 */
EditTickerOwnerTxData.fromRlp = function fromRlp(data) {
  return EditTickerOwnerTxData.fromBufferFields(new minterjsTx.TxDataEditTickerOwner(data));
};

module.exports = EditTickerOwnerTxData;
