// minter-js-sdk/dist/cjs/tx-data/pool-sell-all.js
'use strict';

var minterjsTx = require('minterjs-tx');
var minterjsUtil = require('minterjs-util');
var utils = require('../utils.js');

/**
 * @param {object} txData
 * @param {Array<number|string>} txData.coins - list of coin id
 * @param {number|string} [txData.minimumValueToBuy=0]
 * @param {TxOptions} [options]
 * @constructor
 */
function SellAllPoolTxData(_ref) {
  var coins = _ref.coins,
    _ref$minimumValueToBu = _ref.minimumValueToBuy,
    minimumValueToBuy = _ref$minimumValueToBu === void 0 ? 0 : _ref$minimumValueToBu;
  var options = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
  if (!options.disableValidation) {
    utils.validateUintArray(coins, 'coins');
    utils.validateAmount(minimumValueToBuy, 'minimumValueToBuy');
  }
  this.coins = coins;
  this.minimumValueToBuy = minimumValueToBuy;
  this.txData = new minterjsTx.TxDataSellAllSwapPool({
    coins: coins.map(function (coin) {
      return utils.integerToHexString(coin);
    }),
    minimumValueToBuy: "0x".concat(minterjsUtil.convertToPip(minimumValueToBuy, 'hex'))
  });
  utils.proxyNestedTxData(this);
}

/**
 * @param {object} txData
 * @param {Array<Buffer>} txData.coins
 * @param {Buffer|string} txData.minimumValueToBuy
 * @param {TxOptions} [options]
 * @return {SellAllPoolTxData}
 */
SellAllPoolTxData.fromBufferFields = function fromBufferFields(_ref2) {
  var coins = _ref2.coins,
    minimumValueToBuy = _ref2.minimumValueToBuy;
  var options = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
  return new SellAllPoolTxData({
    coins: coins.map(function (item) {
      return utils.bufferToInteger(item);
    }),
    minimumValueToBuy: utils.dataPipToAmount(minimumValueToBuy)
  }, options);
};

/**
 * @param {Buffer|string} data
 * @return {SellAllPoolTxData}
 */
SellAllPoolTxData.fromRlp = function fromRlp(data) {
  return SellAllPoolTxData.fromBufferFields(new minterjsTx.TxDataSellAllSwapPool(data));
};

module.exports = SellAllPoolTxData;
