// minter-js-sdk/dist/cjs/tx-data/send.js
'use strict';

var minterjsTx = require('minterjs-tx');
var minterjsUtil = require('minterjs-util');
var utils = require('../utils.js');

/**
 * @param {object} txData
 * @param {string} txData.to
 * @param {number|string} txData.value
 * @param {number|string} txData.coin - coin id
 * @param {TxOptions} [options]
 * @constructor
 */
function SendTxData(_ref) {
  var to = _ref.to,
    _ref$value = _ref.value,
    value = _ref$value === void 0 ? 0 : _ref$value,
    coin = _ref.coin;
  var options = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
  if (!options.disableValidation) {
    utils.validateAddress(to, 'to');
    utils.validateUint(coin, 'coin');
    utils.validateAmount(value, 'value');
  }
  this.to = to;
  this.value = value;
  this.coin = coin;
  this.txData = new minterjsTx.TxDataSend({
    to: minterjsUtil.toBuffer(to),
    coin: utils.integerToHexString(coin),
    value: "0x".concat(minterjsUtil.convertToPip(value, 'hex'))
  });
  utils.proxyNestedTxData(this);
}

/**
 * @param {object} txData
 * @param {Buffer|string} txData.to
 * @param {Buffer|string|number} txData.value
 * @param {Buffer|string|number} txData.coin
 * @param {TxOptions} [options]
 * @return {SendTxData}
 */
SendTxData.fromBufferFields = function fromBufferFields(_ref2) {
  var to = _ref2.to,
    value = _ref2.value,
    coin = _ref2.coin;
  var options = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
  return new SendTxData({
    to: utils.dataToAddress(to),
    coin: utils.dataToInteger(coin),
    value: utils.dataPipToAmount(value)
  }, options);
};

/**
 * @param {Buffer|string} data
 * @return {SendTxData}
 */
SendTxData.fromRlp = function fromRlp(data) {
  return SendTxData.fromBufferFields(new minterjsTx.TxDataSend(data));
};

module.exports = SendTxData;
