// minter-js-sdk/dist/cjs/tx-data/stake-unbond.js
'use strict';

var minterjsTx = require('minterjs-tx');
var minterjsUtil = require('minterjs-util');
var utils = require('../utils.js');

/**
 * @param {object} txData
 * @param {string} txData.publicKey
 * @param {number|string} txData.coin - coin id
 * @param {number|string} txData.stake
 * @param {TxOptions} [options]
 * @constructor
 */
function UnbondTxData(_ref) {
  var publicKey = _ref.publicKey,
    coin = _ref.coin,
    stake = _ref.stake;
  var options = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
  if (!options.disableValidation) {
    utils.validatePublicKey(publicKey, 'publicKey');
    utils.validateUint(coin, 'coin');
    utils.validateAmount(stake, 'stake');
  }
  this.publicKey = publicKey;
  this.coin = coin;
  this.stake = stake;
  this.txData = new minterjsTx.TxDataUnbond({
    publicKey: minterjsUtil.toBuffer(publicKey),
    coin: utils.integerToHexString(coin),
    stake: "0x".concat(minterjsUtil.convertToPip(stake, 'hex'))
  });
  utils.proxyNestedTxData(this);
}

/**
 * @param {object} txData
 * @param {Buffer|string} txData.publicKey
 * @param {Buffer|string} txData.stake
 * @param {Buffer|string} txData.coin
 * @param {TxOptions} [options]
 * @return {UnbondTxData}
 */
UnbondTxData.fromBufferFields = function fromBufferFields(_ref2) {
  var publicKey = _ref2.publicKey,
    coin = _ref2.coin,
    stake = _ref2.stake;
  var options = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
  return new UnbondTxData({
    publicKey: utils.dataToPublicKey(publicKey),
    coin: utils.dataToInteger(coin),
    stake: utils.dataPipToAmount(stake)
  }, options);
};

/**
 * @param {Buffer|string} data
 * @return {UnbondTxData}
 */
UnbondTxData.fromRlp = function fromRlp(data) {
  return UnbondTxData.fromBufferFields(new minterjsTx.TxDataUnbond(data));
};

module.exports = UnbondTxData;
