// minter-js-sdk/dist/cjs/tx-data/vote-halt-block.js
'use strict';

var minterjsTx = require('minterjs-tx');
var minterjsUtil = require('minterjs-util');
var utils = require('../utils.js');

/**
 * @param {object} txData
 * @param {string} txData.publicKey
 * @param {number|string} txData.height
 * @param {TxOptions} [options]
 * @constructor
 */
function VoteHaltBlockTxData(_ref) {
  var publicKey = _ref.publicKey,
    height = _ref.height;
  var options = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
  if (!options.disableValidation) {
    utils.validatePublicKey(publicKey, 'publicKey');
    utils.validateUint(height, 'height');
  }
  this.publicKey = publicKey;
  this.height = height;
  this.txData = new minterjsTx.TxDataSetHaltBlock({
    publicKey: minterjsUtil.toBuffer(publicKey),
    height: utils.integerToHexString(height)
  });
  utils.proxyNestedTxData(this);
}

/**
 * @param {object} txData
 * @param {Buffer|string} txData.publicKey
 * @param {Buffer|string|number} txData.height
 * @param {TxOptions} [options]
 * @return {VoteHaltBlockTxData}
 */
VoteHaltBlockTxData.fromBufferFields = function fromBufferFields(_ref2) {
  var publicKey = _ref2.publicKey,
    height = _ref2.height;
  var options = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
  return new VoteHaltBlockTxData({
    publicKey: utils.dataToPublicKey(publicKey),
    height: utils.dataToInteger(height)
  }, options);
};

/**
 * @param {Buffer|string} data
 * @return {VoteHaltBlockTxData}
 */
VoteHaltBlockTxData.fromRlp = function fromRlp(data) {
  return VoteHaltBlockTxData.fromBufferFields(new minterjsTx.TxDataSetHaltBlock(data));
};

module.exports = VoteHaltBlockTxData;
